src/core/public/application/application_service.tsx

```diff
@@ -257,6 +260,16 @@ export class ApplicationService {
       const currentAppId = this.currentAppId$.value;
       const navigatingToSameApp = currentAppId === appId;
       const shouldNavigate = navigatingToSameApp ? true : await this.shouldNavigate(overlays);
+      const targetApp = applications$.value.get(appId);
+      if (
+        workspaces.currentWorkspaceId$.value &&
+        targetApp?.visibility === AppVisibility.homeOnly
+      ) {
+        // If user is inside a workspace and the target app is homeOnly
+        // refresh the page by doing a hard navigation
+        window.location.assign(getAppUrl(availableMounters, appId, path));
+        return;
+      }
```

Comment:
Should this navigation logic moved to `if (shouldNavigate) {}` clause?

Reply:
Agree, done for that.